google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/apis/CloudLibraryDependencyWriter.java

```diff
@@ -121,23 +130,81 @@ protected void run(@NotNull Result result) throws Throwable {
         // user when it's complete.
         List<MavenId> newMavenIds = mavenIdsMap.get(false);
         if (!newMavenIds.isEmpty()) {
-          newMavenIds.forEach(
-              mavenId -> {
-                MavenDomUtil.createDomDependency(model, /* editor= */ null, mavenId);
-                UsageTrackerProvider.getInstance()
-                    .trackEvent(GctTracking.CLIENT_LIBRARY_ADD_LIBRARY)
-                    .addMetadata(
-                        GctTracking.METADATA_BUILD_SYSTEM_KEY,
-                        GctTracking.METADATA_BUILD_SYSTEM_MAVEN)
-                    .addMetadata(GctTracking.METADATA_LABEL_KEY, mavenId.getDisplayString())
-                    .ping();
-              });
+          newMavenIds.forEach(mavenId -> writeNewMavenDependency(model, mavenId, bomVersion));
+
+          if (bomVersion != null) {
+            addBomToMavenModule(model, bomVersion);
+          }
+
           notifyAddedDependencies(newMavenIds, project);
         }
       }
     }.execute();
   }
 
+  /**
+   * Adds the google-cloud-java BOM dependency to the module's dependencyManagement section.
+   *
+   * <p>If the BOM already exists, then this will update the version of the BOM to match the
+   * supplied version.
+   */
+  private static void addBomToMavenModule(MavenDomProjectModel model, String bomVersion) {
+    MavenDomDependencies mavenDomDependencies = model.getDependencyManagement().getDependencies();
+    Optional<MavenDomDependency> bomDependencyOptional =
+        mavenDomDependencies
+            .getDependencies()
+            .stream()
+            .filter(
+                mdd ->
+                    CloudApiMavenService.GOOGLE_CLOUD_JAVA_BOM_ARTIFACT.equals(
+                        mdd.getArtifactId().getStringValue()))
+            .findFirst();
+
+    if (!bomDependencyOptional.isPresent()) {
+      writeNewBom(model, bomVersion);
+    } else {
+      // update the version
+      bomDependencyOptional.get().getVersion().setStringValue(bomVersion);
+    }
+  }
+
+  /**
+   * Writes the google-cloud-java dependency to the users pom.xml.
+   *
+   * <p>If a BOM version is supplied, the dependency will not be written with a version so that the
+   * BOM can manage the version.
+   */
+  private static void writeNewMavenDependency(
+      MavenDomProjectModel model, MavenId mavenId, @Nullable String bomVersion) {
+    MavenDomDependency dependency = MavenDomUtil.createDomDependency(model, /* editor= */ null);
+    dependency.getGroupId().setStringValue(mavenId.getGroupId());
+    dependency.getArtifactId().setStringValue(mavenId.getArtifactId());
+
+    // Only write the version to the dependency if there is no BOM
+    if (bomVersion == null) {
+      dependency.getVersion().setStringValue(mavenId.getVersion());
```

Comment:
So just for refreshing on this, once `createDomDependency` is called, calling the `set` method family updates `pom.xml` automatically?

Reply:
correct. 

The previously used overloaded `createDomDependency` method takes in a MavenId and does these `setStringValue` calls for you. But since we now want to only conditionally set the version, we have to do these externally.